scala_proto/private/scalapb_aspect.bzl

```diff
@@ -0,0 +1,235 @@
+        output_jar = output,
+        compile_jar = output,
+    )
+
+def _empty_java_info(deps_java_info, implicit_deps):
+    return java_common.merge(deps_java_info + implicit_deps)
+
+####
+# This is applied to the DAG of proto_librarys reachable from a deps
+# or a scalapb_scala_library. Each proto_library will be one scalapb
+# invocation assuming it has some sources.
+def _scalapb_aspect_impl(target, ctx):
+    deps = [d[ScalaPBAspectInfo].java_info for d in ctx.rule.attr.deps]
+
+    if ProtoInfo not in target:
+        # We allow some dependencies which are not protobuf, but instead
+        # are jvm deps. This is to enable cases of custom generators which
+        # add a needed jvm dependency.
+        java_info = target[JavaInfo]
+        src_jars = depset()
+        outs = depset()
+        transitive_ti = merge_proto_infos(
+            [
+                d[ScalaPBAspectInfo].proto_info
+                for d in ctx.rule.attr.deps
+            ],
+        )
+    else:
+        target_ti = target[ProtoInfo]
+        transitive_ti = merge_proto_infos(
+            [
+                d[ScalaPBAspectInfo].proto_info
+                for d in ctx.rule.attr.deps
+            ] + [target_ti],
+        )
+        # we sort so the inputs are always the same for caching
+        compile_protos = sorted(target_ti.direct_sources)
+        transitive_protos = sorted(target_ti.transitive_sources)
+
+        toolchain = ctx.toolchains["@io_bazel_rules_scala//scala_proto:toolchain_type"]
+        flags = []
+        imps = [j[JavaInfo] for j in toolchain.implicit_compile_deps]
+
+        if toolchain.with_grpc:
+            flags.append("grpc")
+            imps.extend([j[JavaInfo] for j in toolchain.grpc_deps])
+
+        if toolchain.with_flat_package:
+            flags.append("flat_package")
+
+        if toolchain.with_single_line_to_string:
+            flags.append("single_line_to_proto_string")
+
+        # This feels rather hacky and odd, but we can't compare the labels to ignore a target easily
+        # since the @ or // forms seem to not have good equality :( , so we aim to make them absolute
+        #
+        # the backlisted protos in the tool chain get made into to absolute paths
+        # so we make the local target we are looking at absolute too
+        target_absolute_label = target.label
+        if not str(target_absolute_label)[0] == "@":
+            target_absolute_label = Label("@%s//%s:%s" % (ctx.workspace_name, target.label.package, target.label.name))
+
+        for lbl in toolchain.blacklisted_protos:
+            if(lbl.label == target_absolute_label):
+                compile_protos = False
+
+        code_generator = toolchain.code_generator
+        for lbl in toolchain.override_code_generator_targets:
+            if(lbl.label == target_absolute_label):
+                code_generator = toolchain.override_code_generator
+
+
+        for lbl in toolchain.override_code_generator_targets:
```

Comment:
This is identical to three lines above. I guess it should be removed.

Reply:
👍